### ISSUE TITLE
Datagrid disrupting focus when destroyed

### ISSUE THREAD

**Issue by jzabinski** [vmware/clarity#3840]
## Bug Description

_Prerequisites_

1. clr-dg-datagrid must be used by a component
2. clr-dg-pagesize must be used as a child component of requirement #1's clr-dg-datagrid

_Description_
If an element outside the datagrid currently has focus, that element will lose focus if the datagrid is removed from the page.

## How to reproduce

See [this Stackblitz example](https://stackblitz.com/edit/angular-myr5dy). The example displays a Clarity datagrid with a child component that displays the page size. The datagrid is housed in a div like this:
`<div *ngIf="knights.length > 0"><clr-datagrid> [more code]</clr-datagrid></div>`
The intention is for the datagrid to only display if there are rows to display.

It also has an input element with an id defined as 'example', and a button that calls this function:
`onBtnClick() {
    document.getElementById('example').focus();
    this.knights = new Array<Info>();
  }`

1.  Click on the button 'Click this'
2.  Note the function being called executes and the array holding information for the datagrid is reset. Angular correctly detects this change and removes the datagrid, but the input does not have focus.

## Expected behavior

The focus of elements should be unaffected by the removal of Clarity components from the page.

## Versions

**App**

* Angular: 8.2.7
* Node: 8.9.4
* Clarity: 2.2.0

**Device:**

* Type: HP Elitebook
* OS: Windows 10
* Browser Chrome
* Version 76

## Additional notes

The root cause appears to be that, when the child component for the page size is removed, then this code is called:
```
datagrid-pagination.ts:
ngOnDestroy() {
    --> this.page.resetPageSize();
    if (this._pageSubscription) {
      this._pageSubscription.unsubscribe();
    }
  }
```
And the datagrid is listening for changes to the pagesize, where it sets focus to itself if there are any changes to the page size:
```
datagrid.ts:
this.page.change.subscribe(() => {
        this.datagridTable.nativeElement.focus();
      }),
```
The issue is that (1) some other element had focus at the time, and (2) the entire datagrid is being destroyed anyway, and so the datagrid should be unconcerned with whether it has focus. 

On the one hand, it could be argued that this could be considered an iteration of [the longstanding need in Angular for ngBeforeDestroy](https://github.com/angular/angular/issues/17404). If an ngBeforeDestroy were to be called before the Clarity components are destroyed, then the datagrid can ignore questions of where to focus since doom is nigh.
On the other hand, it could be argued that the intention of focusing on the datagrid when the datagrid's child components are destroyed assumes that focus was on the child component in the first place. In the case of the removal of the datagrid entirely, this is an invalid assumption. Maybe the Clarity element that currently has focus could be tracked in a service, and the datagrid would only take focus when one of its children lost focus. This is a lot uglier than waiting for OnBeforeDestroy though.

**Comment by Antony74** [comment]
Note that this bug can also cause unwanted scrolling.  However there's not enough on the page of the StackBlitz example to manifest this additional symptom.

There is a general purpose workaround for the lack of an ngBeforeDestroy.  Insert a child component and it's guaranteed to be destroyed first, therefore it can send a warning to its parent.  e.g.

``` html
<on-destroy-canary (destroyWarning)="ngBeforeDestroy()"></iw-on-destroy-canary>
```

Which I've implemented thus

``` TypeScript
import { Component, EventEmitter, OnDestroy, Output } from '@angular/core';

@Component({
    selector: 'on-destroy-canary',
    template: ''
  })
export class OnDestroyCanaryComponent implements OnDestroy {

  @Output() destroyWarning = new EventEmitter<{}>();

  ngOnDestroy() {
    // We have been destroyed.  Therefore our parent is about to be
    // destroyed (because we never put attribute-conditionals
    // such as *ngIf on an <on-destroy-canary>, that would defeat
    // the purpose).  Warn the parent of it's impending fate
    // (your canary has died, you need to leave the mine now!)
    this.destroyWarning.emit();
  }

}
```

In this case the inserted component is more difficult to place correctly because we need to ensure the inserted component is positioned correctly so that is it destroyed before the pagination control, not just the datagrid itself.

I'll continue so as to present a complete workaround for the present bug, but I'm not at all proud of what I do when I receive the manufactured ngBeforeDestroy event.   You're welcome to tell me this is very ugly, and even more welcome to tell me how it could be improved upon, but it does prevent the unwanted focus( ) call from having any effect for want of a visible target to focus upon.

``` TypeScript
  constructor(private element: ElementRef) {
    super();
  }

  hideRecursive(elm) {
    elm.style.display = 'none';
    for (const child of elm.children) {
      this.hideRecursive(child);
    }
  }

  ngBeforeDestroy() {
    // We're about to be destroyed.  It would be a shame if focus() was called on a child
    // element, especially if this caused us to scroll to an element that is about to be
    // destroyed (https://github.com/vmware/clarity/issues/3840).  This is very crude, but
    // we hide by setting the 'display: none' style on ourself and recursively all
    // children, so calling focus() on these elements will have no effect as there won't
    // be anything to focus on.
    this.hideRecursive(this.element.nativeElement);
  }
```

**Comment by jzabinski** [comment]
Can you elaborate on how this can cause unwanted scrolling? I made [a new Stackblitz example](https://angular-mhdqqn.stackblitz.io), and I don't see any scrolling happen, but obviously there are a lot of elements out there, so you might know something that I don't.

The workaround that I ended up using follows the same principle as yours, where the child element notifies the parent element that OnDestroy was called via an EventEmitter. It just so happens that the child element exclusively houses the clr-datagrid too, so that when the datagrid goes away, that is the only thing affected. Essentially the datagrid doubles as the canary. Like so:

`<div *ngIf="knights.length > 0"> <my-datagrid [datagridKnights]="knights" (destroyed)="refocus()" ></my-datagrid> </div>`

Note that if we were to house the containing div in my-datagrid, my-datagrid is actually not destroyed between Angular lifecycles: the *ngIf is an important part of the construct here.

**Comment by Antony74** [comment]
umm... you don't see unwanted scrolling in your latest Stackblitz example because you've already applied a workaround to it?

``` html
<input id="example">
<button (click)="onBtnClick()">Click this</button>

<img src="https://upload.wikimedia.org/wikipedia/commons/f/f9/Phoenicopterus_ruber_in_S%C3%A3o_Paulo_Zoo.jpg">

<div *ngIf="knights.length > 0">
<my-datagrid [datagridKnights]="knights" (destroyed)="refocus()" ></my-datagrid>
</div>
```

If you remove the 'destroyed' property from <my-datagrid> you should see the unwanted scrolling behavior restored.  Thank you very much for reproducing my bug (or very nearly - easy mistake to make!), I'm relieved it didn't turn out to be more obscure.

**Comment by Shijir** [comment]
This bug has been fixed in https://github.com/vmware/clarity/pull/3872. So you will get the fix if you upgrade your Clarity version to 2.3.0: https://clarity.design/news/2.3.0
